Model Element reports extra memory cost to JSC

amarchenko opened the pull request:
#### 910c6466a03f7b130f5ecb1119bbd15a188ae963
<pre>
Model Element reports extra memory cost to JSC
<a href="https://bugs.webkit.org/show_bug.cgi?id=293150">https://bugs.webkit.org/show_bug.cgi?id=293150</a>
<a href="https://rdar.apple.com/143763757">rdar://143763757</a>

Reviewed by Ada Chan and Keith Miller.

Extra memory use is calculated based on resources loaded into the
HTMLModelElement (model data and, if available, environment data).
Additionally, memory cost members are atomic since they could be called
from GC threads via memoryCost().
Overall, this change should make GC more aggressive on the Model Element
and help with cases like page reloading.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::didFinishLoading):
(WebCore::HTMLModelElement::didFailLoading):
(WebCore::HTMLModelElement::didFinishEnvironmentMapLoading):
(WebCore::HTMLModelElement::environmentMapResourceFinished):
(WebCore::HTMLModelElement::modelResourceFinished):
(WebCore::HTMLModelElement::reportExtraMemoryCost):
(WebCore::HTMLModelElement::memoryCost const):
(WebCore::HTMLModelElement::externalMemoryCost const):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:

Canonical link: <a href="https://commits.webkit.org/295220@main">https://commits.webkit.org/295220@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/107a82a1fdc49f04addd9540e3f68d00609cf167

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104388 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14508 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109590 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55056 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106428 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24469 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32639 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79257 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107394 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19019 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94206 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59587 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18828 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12281 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54418 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88513 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12333 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111974 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23236 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88295 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31909 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90450 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87961 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22412 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32860 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10626 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/26013 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31473 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36798 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31267 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34603 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32827 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1336,6 +1348,40 @@ void HTMLModelElement::removedFromAncestor(RemovalType removalType, ContainerNod
+            return;
+        JSC::VM& vm = context->vm();
+        JSC::JSLockHolder lock(vm);
+        vm.heap.reportExtraMemoryAllocated(static_cast<JSCell*>(nullptr), currentCost);
```
Shouldn't this be `currentCost - m_reportedDataMemoryCost`?

You also shouldn't need the `static_cast<JSCell*>` since nullptr is a subtype of all pointer types in C++.

I think this is going to double count your memory between `CodeGeneratorJS.pm`'s `finishCreation` and this?

amarchenko replied:
Changed to `currentCost - m_reportedDataMemoryCost` and `nullptr`.
Could you please elaborate on CodeGeneratorJS.pm's finishCreation in this context?

kmiller68 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1050,8 +1057,11 @@ void HTMLModelElement::environmentMapResourceFinished()
-    if (m_modelPlayer)
+    if (m_modelPlayer) {
+        if (m_environmentMapData.size() > m_environmentMapDataMemoryCost.load(std::memory_order_relaxed))
+            m_environmentMapDataMemoryCost.store(m_environmentMapData.size(), std::memory_order_relaxed);
```
Isn't there an ABA problem here? I think you want a `fetch_max`.

amarchenko replied:
Unfortunately, `fetch_max` is not available in C++23.
Since the value is updated from a single thread, there shouldn't be ABA here.

kmiller68 replied:
If these aren't concurrent callers of this why not just report the delta of extra memory here?

amarchenko replied:
This is an intermediate step, the data is passed further to the Model Process to validate and load in REKit. When that is done, `HTMLModelElement::didFinishEnvironmentMapLoading(true)` is called, that's when we have confidence to report to JSC.

kmiller68 replied:
Oh I see. If failing is rare then it's probably not a big deal to report eagerly, the worst case scenario is we do a GC slightly early.

kmiller68 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1082,6 +1092,8 @@ void HTMLModelElement::modelResourceFinished()
 
     m_dataComplete = true;
+    if (m_data.size() > m_dataMemoryCost.load(std::memory_order_relaxed))
+        m_dataMemoryCost.store(m_data.size(), std::memory_order_relaxed);
```
Ditto.

amarchenko replied:
Same as above.

kmiller68 replied:
Ditto.

amarchenko replied:
Similar to above with a difference of `HTMLModelElement::didFinishLoading` is called,

kmiller68 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1336,6 +1348,40 @@ void HTMLModelElement::removedFromAncestor(RemovalType removalType, ContainerNod
+        if (!context)
+            return;
+        JSC::VM& vm = context->vm();
+        JSC::JSLockHolder lock(vm);
```
Do you need to take the API lock here? Does it crash if you replace this with `ASSERT(vm.currentThreadIsHoldingAPILock())`?

amarchenko replied:
The ASSERT doesn't crash.
Also, added `ASSERT_WITH_MESSAGE` per offline discussion.

kmiller68 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1336,6 +1348,41 @@ void HTMLModelElement::removedFromAncestor(RemovalType removalType, ContainerNod
     }
 }
 
+void HTMLModelElement::reportExtraMemoryCost()
```
If you reported the extra memory in the above functions I don't think you need this function at all.

amarchenko replied:
Reporting extra memory in the functions above is too early, the data doesn't stay here (notice `m_data.takeAsContiguous()`) there are further steps that might fail (in the Model Process). That's why we report only on successful callbacks.

kmiller68 reviewed (approved):
r=me on the JSC side. Not sure if someone else should review for the correctness of size calculations.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -621,10 +625,13 @@ void HTMLModelElement::didUpdateBoundingBox(ModelPlayer&, const FloatPoint3D& ce
-        else
+        } else {
             m_environmentMapReadyPromise->reject(Exception { ExceptionCode::AbortError });
+            m_environmentMapDataMemoryCost.store(0, std::memory_order_relaxed);
```
Should we set `m_environmentMapDataMemoryCost` to `0` in `HTMLModelElement::setEnvironmentMap()`?

If there's an update to the `environmentmap` URL, it can be set to `0` right away. If the `environmentmap` is being set to empty, then call `reportExtraMemoryCost()` right away.

Otherwise, in `HTMLModelElement::environmentMapResourceFinished()`, `m_environmentMapDataMemoryCost` can be updated to the size of the data.

Then in `HTMLModelElement::didFinishEnvironmentMapLoading()`, `m_environmentMapDataMemoryCost` is reset to `0` for the error case. For all cases, call `reportExtraMemoryCost()`  to report the final memory cost.

amarchenko replied:
Made changes to address the described workflow for `m_environmentMapDataMemoryCost`.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -515,6 +517,8 @@ void HTMLModelElement::didFailLoading(ModelPlayer& modelPlayer, const ResourceEr
     if (!m_readyPromise->isFulfilled())
         m_readyPromise->reject(Exception { ExceptionCode::AbortError });
+
+    m_dataMemoryCost.store(0, std::memory_order_relaxed);
```
Should we set `m_dataMemoryCost` to 0 in `HTMLModelElement::setSourceURL()`?

If there's an update to the source URL, the cost can be set to 0 right away. If the source is being set to empty, then call reportExtraMemoryCost() right away.

Otherwise, in `HTMLModelElement::modelResourceFinished()`, `m_dataMemoryCost` can be updated to the size of the data.

Then in `HTMLModelElement::didFailLoading()`,`m_dataMemoryCost` is reset to 0 and `reportExtraMemoryCost()` should be called to report the final memory cost. In `HTMLModelElement::didFinishLoading()`, `reportExtraMemoryCost()` should be called to report the final memory cost.

amarchenko replied:
Made changes to address the described workflow for `m_dataMemoryCost`.

amarchenko committed:
Model Element reports extra memory cost to JSC
https://bugs.webkit.org/show_bug.cgi?id=293150
rdar://143763757

Reviewed by Ada Chan and Keith Miller.

Extra memory use is calculated based on resources loaded into the
HTMLModelElement (model data and, if available, environment data).
Additionally, memory cost members are atomic since they could be called
from GC threads via memoryCost().
Overall, this change should make GC more aggressive on the Model Element
and help with cases like page reloading.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::didFinishLoading):
(WebCore::HTMLModelElement::didFailLoading):
(WebCore::HTMLModelElement::didFinishEnvironmentMapLoading):
(WebCore::HTMLModelElement::environmentMapResourceFinished):
(WebCore::HTMLModelElement::modelResourceFinished):
(WebCore::HTMLModelElement::reportExtraMemoryCost):
(WebCore::HTMLModelElement::memoryCost const):
(WebCore::HTMLModelElement::externalMemoryCost const):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/HTMLModelElement.idl:

Canonical link: https://commits.webkit.org/295220@main
